File: apps/bff/src/modules/task/adapters/sql.ts
import type {
  CreateProjectInput,
  CreateTaskInput,
  ListChangesInput,
  ListTasksByProjectInput,
  ListTasksRangeInput,
  ProjectStore,
  ProjectTaskCounts,
  SyncReadStore,
  TaskRefInput,
  TaskStore,
  UpdateTaskInput,
} from '@bff/modules/task/ports';
import type { DatabaseSchema } from '@bff/platform/db/schema';
import { currentSyncSeq, withUserSyncSeq } from '@bff/platform/db/sync';
import { newId } from '@bff/platform/ids';
import { orderKeyAfter } from '@bff/platform/ordering';
import type { Project, ProjectChange, SyncChangesResponse, Task, TaskChange, TaskRangeResponse } from '@tooday/shared';
import type { Kysely } from 'kysely';

const TASK_COLUMNS = ['id', 'project_id', 'title', 'date', 'start_at', 'duration_min', 'status', 'version'] as const;
const TASK_CHANGE_COLUMNS = [...TASK_COLUMNS, 'sync_seq', 'deleted_at'] as const;

interface TaskRow {
  id: string;
  project_id: string | null;
  title: string;
  date: string;
  start_at: string;
  duration_min: number;
  status: Task['status'];
  version: number;
}

function toTask(row: TaskRow): Task {
  return {
    id: row.id,
    projectId: row.project_id,
    title: row.title,
    date: row.date,
    startAt: row.start_at,
    durationMin: row.duration_min,
    status: row.status,
    version: row.version,
  };
}

function toTaskChange(row: TaskRow & { sync_seq: number; deleted_at: Date | null }): TaskChange {
  return { ...toTask(row), syncSeq: row.sync_seq, deleted: row.deleted_at !== null };
}

/*
 * 낱개 SELECT는 모듈 함수로 둔다 — 스토어 메서드와 SqlSyncReadStore의 스냅샷
 * 트랜잭션이 같은 쿼리를 공유하기 위해서다 (Transaction은 Kysely에 대입 가능).
 */

function selectProjects(db: Kysely<DatabaseSchema>, userId: string): Promise<Project[]> {
  return db
    .selectFrom('projects')
    .select(['id', 'name', 'color'])
    .where('user_id', '=', userId)
    .where('deleted_at', 'is', null)
    .orderBy('position')
    .orderBy('id')
    .execute();
}

async function selectTasksInRange(db: Kysely<DatabaseSchema>, { userId, from, to }: ListTasksRangeInput): Promise<Task[]> {
  const rows = await db
    .selectFrom('tasks')
    .select(TASK_COLUMNS)
    .where('user_id', '=', userId)
    .where('deleted_at', 'is', null)
    .where('date', '>=', from)
    .where('date', '<=', to)
    .orderBy('date')
    .orderBy('start_at')
    .execute();
  return rows.map(toTask);
}

async function selectTaskChanges(db: Kysely<DatabaseSchema>, { userId, cursor }: ListChangesInput): Promise<TaskChange[]> {
  const rows = await db
    .selectFrom('tasks')
    .select(TASK_CHANGE_COLUMNS)
    .where('user_id', '=', userId)
    .where('sync_seq', '>', cursor)
    .orderBy('sync_seq')
    .execute();
  return rows.map(toTaskChange);
}

async function selectProjectChanges(db: Kysely<DatabaseSchema>, { userId, cursor }: ListChangesInput): Promise<ProjectChange[]> {
  const rows = await db
    .selectFrom('projects')
    .select(['id', 'name', 'color', 'sync_seq', 'deleted_at'])
    .where('user_id', '=', userId)
    .where('sync_seq', '>', cursor)
    .orderBy('sync_seq')
    .execute();
  return rows.map((row) => ({
    id: row.id,
    name: row.name,
    color: row.color,
    syncSeq: row.sync_seq,
    deleted: row.deleted_at !== null,
  }));
}

/**
 * 풀드 Postgres에서는 쿼리마다 커넥션(=READ COMMITTED 스냅샷)이 다르므로, 데이터와
 * 커서를 REPEATABLE READ 트랜잭션 하나(단일 스냅샷)에서 읽는다. 쓰기는 유저 advisory
 * lock으로 직렬화되어 seq 발급 순서 = 커밋 순서이고 행과 카운터는 같은 트랜잭션에서
 * 커밋되므로, 스냅샷의 counters.seq가 N이면 seq ≤ N인 변경 전부가 그 스냅샷에 보인다
 * — 커서를 counters에서 도출하면 "응답에 없는 변경을 커서가 지나치는" 구멍이 없다.
 */
export class SqlSyncReadStore implements SyncReadStore {
  constructor(private readonly db: Kysely<DatabaseSchema>) {}

  async range({ userId, from, to }: ListTasksRangeInput): Promise<TaskRangeResponse> {
    return this.db
      .transaction()
      .setIsolationLevel('repeatable read')
      .execute(async (trx) => {
        const cursor = await currentSyncSeq(trx, userId);
        const tasks = await selectTasksInRange(trx, { userId, from, to });
        const projects = await selectProjects(trx, userId);
        return { tasks, projects, cursor };
      });
  }

  async changes({ userId, cursor }: ListChangesInput): Promise<SyncChangesResponse> {
    return this.db
      .transaction()
      .setIsolationLevel('repeatable read')
      .execute(async (trx) => {
        // 카운터에는 행 없는 쓰기(not-found update/remove)의 seq 갭이 있을 수 있어
        // 반환 행의 max보다 앞설 수 있다 — 그래도 스냅샷에 보인 변경은 전부 실렸으므로 안전
        const next = await currentSyncSeq(trx, userId);
        const tasks = await selectTaskChanges(trx, { userId, cursor });
        const projects = await selectProjectChanges(trx, { userId, cursor });
        return { tasks, projects, cursor: Math.max(cursor, next) };
      });
  }
}

export class SqlProjectStore implements ProjectStore {
  constructor(private readonly db: Kysely<DatabaseSchema>) {}

  async listByUser(userId: string): Promise<Project[]> {
    return selectProjects(this.db, userId);
  }

  async findById({ userId, id }: { userId: string; id: string }): Promise<Project | null> {
    const row = await this.db
      .selectFrom('projects')
      .select(['id', 'name', 'color'])
      .where('user_id', '=', userId)
      .where('id', '=', id)
      .where('deleted_at', 'is', null)
      .executeTakeFirst();
    return row ?? null;
  }

  async create({ userId, name, color }: CreateProjectInput): Promise<Project> {
    return withUserSyncSeq(this.db, userId, async (trx, seq) => {
      const last = await trx
        .selectFrom('projects')
        .select('position')
        .where('user_id', '=', userId)
        .orderBy('position', 'desc')
        .limit(1)
        .executeTakeFirst();
      const project: Project = { id: newId(), name, color };
      await trx
        .insertInto('projects')
        .values({
          ...project,
          user_id: userId,
          position: orderKeyAfter(last?.position ?? null),
          sync_seq: seq,
          deleted_at: null,
        })
        .execute();
      return project;
    });
  }
}

export class SqlTaskStore implements TaskStore {
  constructor(private readonly db: Kysely<DatabaseSchema>) {}

  async findById({ userId, id }: TaskRefInput): Promise<Task | null> {
    const row = await this.db
      .selectFrom('tasks')
      .select(TASK_COLUMNS)
      .where('user_id', '=', userId)
      .where('id', '=', id)
      .where('deleted_at', 'is', null)
      .executeTakeFirst();
    return row ? toTask(row) : null;
  }

  async listByProject({ userId, projectId }: ListTasksByProjectInput): Promise<Task[]> {
    const rows = await this.db
      .selectFrom('tasks')
      .select(TASK_COLUMNS)
      .where('user_id', '=', userId)
      .where('project_id', '=', projectId)
      .where('deleted_at', 'is', null)
      .orderBy('date')
      .orderBy('start_at')
      .execute();
    return rows.map(toTask);
  }

  async countsByProject(userId: string): Promise<ProjectTaskCounts[]> {
    const rows = await this.db
      .selectFrom('tasks')
      .select((eb) => [
        'project_id',
        eb.fn.countAll().as('total'),
        // COUNT은 NULL을 세지 않으므로 done 행만 1로 남겨 센다
        eb.fn.count(eb.case().when('status', '=', 'done').then(eb.val(1)).end()).as('done'),
      ])
      .where('user_id', '=', userId)
      .where('deleted_at', 'is', null)
      .where('project_id', 'is not', null)
      .groupBy('project_id')
      .execute();
    // pg의 count는 bigint를 문자열로 돌려주므로 여기서 number로 정규화한다
    return rows.map((row) => ({
      projectId: row.project_id as string,
      total: Number(row.total),
      done: Number(row.done),
    }));
  }

  async create({ userId, title, projectId, date, startAt, durationMin }: CreateTaskInput): Promise<Task> {
    return withUserSyncSeq(this.db, userId, async (trx, seq) => {
      const last = await trx
        .selectFrom('tasks')
        .select('position')
        .where('user_id', '=', userId)
        .orderBy('position', 'desc')
        .limit(1)
        .executeTakeFirst();
      const row = await trx
        .insertInto('tasks')
        .values({
          id: newId(),
          user_id: userId,
          project_id: projectId,
          title,
          date,
          start_at: startAt,
          duration_min: durationMin,
          status: 'todo',
          position: orderKeyAfter(last?.position ?? null),
          sync_seq: seq,
          deleted_at: null,
          completed_at: null,
        })
        .returning(TASK_COLUMNS)
        .executeTakeFirstOrThrow();
      return toTask(row);
    });
  }

  async update({ userId, id, patch }: UpdateTaskInput): Promise<Task | null> {
    return withUserSyncSeq(this.db, userId, async (trx, seq) => {
      const now = new Date();
      const row = await trx
        .updateTable('tasks')
        .set((eb) => ({
          ...(patch.title !== undefined ? { title: patch.title } : {}),
          ...(patch.projectId !== undefined ? { project_id: patch.projectId } : {}),
          ...(patch.date !== undefined ? { date: patch.date } : {}),
          ...(patch.startAt !== undefined ? { start_at: patch.startAt } : {}),
          ...(patch.durationMin !== undefined ? { duration_min: patch.durationMin } : {}),
          ...(patch.status !== undefined ? { status: patch.status, completed_at: patch.status === 'done' ? now : null } : {}),
          version: eb('version', '+', 1),
          updated_at: now,
          sync_seq: seq,
        }))
        .where('user_id', '=', userId)
        .where('id', '=', id)
        .where('deleted_at', 'is', null)
        .returning(TASK_COLUMNS)
        .executeTakeFirst();
      return row ? toTask(row) : null;
    });
  }

  async remove({ userId, id }: TaskRefInput): Promise<boolean> {
    return withUserSyncSeq(this.db, userId, async (trx, seq) => {
      const now = new Date();
      const row = await trx
        .updateTable('tasks')
        // tombstone도 쓰기다 — version을 올려 "쓰기마다 단조 증가" 불변식을 지킨다
        .set((eb) => ({ deleted_at: now, updated_at: now, sync_seq: seq, version: eb('version', '+', 1) }))
        .where('user_id', '=', userId)
        .where('id', '=', id)
        .where('deleted_at', 'is', null)
        .returning('id')
        .executeTakeFirst();
      return row !== undefined;
    });
  }
}

File: rust/crates/bff/src/modules/task/adapters/sql.rs
use async_trait::async_trait;
use chrono::Utc;
use sqlx::postgres::PgRow;
use sqlx::{PgPool, Row};
use tooday_shared::{
    Patch, Project, ProjectChange, ProjectColor, Task, TaskChange, TaskStatus,
};

use crate::modules::task::ports::{
    CreateProjectInput, CreateTaskInput, ListChangesInput, ListTasksByProjectInput,
    ListTasksRangeInput, ProjectStore, ProjectTaskCounts, TaskRefInput, TaskStore, UpdateTaskInput,
};
use crate::platform::db::sync::{begin_with_sync_seq, current_sync_seq};
use crate::platform::errors::{StoreError, StoreResult};
use crate::platform::ids::new_id;
use crate::platform::ordering::order_key_after;

const TASK_COLUMNS: &str = "id, project_id, title, date, start_at, duration_min, status, version";

fn to_task(row: &PgRow) -> StoreResult<Task> {
    let status: String = row.try_get("status").map_err(StoreError::infrastructure)?;
    let project_id: Option<uuid::Uuid> = row.try_get("project_id").map_err(StoreError::infrastructure)?;
    Ok(Task {
        id: row.try_get::<uuid::Uuid, _>("id").map_err(StoreError::infrastructure)?.to_string(),
        project_id: project_id.map(|id| id.to_string()),
        title: row.try_get("title").map_err(StoreError::infrastructure)?,
        date: row.try_get("date").map_err(StoreError::infrastructure)?,
        start_at: row.try_get("start_at").map_err(StoreError::infrastructure)?,
        duration_min: row.try_get::<i32, _>("duration_min").map_err(StoreError::infrastructure)? as i64,
        status: TaskStatus::from_str(&status)
            .ok_or_else(|| StoreError::Infrastructure(format!("알 수 없는 status: {status}")))?,
        version: row.try_get::<i32, _>("version").map_err(StoreError::infrastructure)? as i64,
    })
}

fn to_project(row: &PgRow) -> StoreResult<Project> {
    let color: String = row.try_get("color").map_err(StoreError::infrastructure)?;
    Ok(Project {
        id: row.try_get::<uuid::Uuid, _>("id").map_err(StoreError::infrastructure)?.to_string(),
        name: row.try_get("name").map_err(StoreError::infrastructure)?,
        color: ProjectColor::from_str(&color)
            .ok_or_else(|| StoreError::Infrastructure(format!("알 수 없는 color: {color}")))?,
    })
}

pub struct SqlProjectStore {
    pool: PgPool,
}

impl SqlProjectStore {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl ProjectStore for SqlProjectStore {
    async fn list_by_user(&self, user_id: &str) -> StoreResult<Vec<Project>> {
        let rows = sqlx::query(
            r#"select id, name, color from projects
               where user_id = $1::uuid and deleted_at is null
               order by position, id"#,
        )
        .bind(user_id)
        .fetch_all(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;
        rows.iter().map(to_project).collect()
    }

    async fn find_by_id(&self, input: &TaskRefInput) -> StoreResult<Option<Project>> {
        let row = sqlx::query(
            r#"select id, name, color from projects
               where user_id = $1::uuid and id = $2::uuid and deleted_at is null"#,
        )
        .bind(&input.user_id)
        .bind(&input.id)
        .fetch_optional(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;
        row.as_ref().map(to_project).transpose()
    }

    async fn create(&self, input: CreateProjectInput) -> StoreResult<Project> {
        let (mut tx, seq) = begin_with_sync_seq(&self.pool, &input.user_id)
            .await
            .map_err(StoreError::infrastructure)?;
        let last: Option<(String,)> = sqlx::query_as(
            "select position from projects where user_id = $1::uuid order by position desc limit 1",
        )
        .bind(&input.user_id)
        .fetch_optional(&mut *tx)
        .await
        .map_err(StoreError::infrastructure)?;

        let project = Project { id: new_id(), name: input.name, color: input.color };
        sqlx::query(
            r#"insert into projects (id, user_id, name, color, position, sync_seq, deleted_at)
               values ($1::uuid, $2::uuid, $3, $4, $5, $6, null)"#,
        )
        .bind(&project.id)
        .bind(&input.user_id)
        .bind(&project.name)
        .bind(project.color.as_str())
        .bind(order_key_after(last.map(|(position,)| position).as_deref()))
        .bind(seq as i32)
        .execute(&mut *tx)
        .await
        .map_err(StoreError::infrastructure)?;
        tx.commit().await.map_err(StoreError::infrastructure)?;
        Ok(project)
    }

    async fn changes_since(&self, input: ListChangesInput) -> StoreResult<Vec<ProjectChange>> {
        let rows = sqlx::query(
            r#"select id, name, color, sync_seq, deleted_at from projects
               where user_id = $1::uuid and sync_seq > $2
               order by sync_seq"#,
        )
        .bind(&input.user_id)
        .bind(input.cursor as i32)
        .fetch_all(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;

        rows.iter()
            .map(|row| {
                let deleted_at: Option<chrono::DateTime<Utc>> =
                    row.try_get("deleted_at").map_err(StoreError::infrastructure)?;
                Ok(ProjectChange {
                    project: to_project(row)?,
                    sync_seq: row.try_get::<i32, _>("sync_seq").map_err(StoreError::infrastructure)? as i64,
                    deleted: deleted_at.is_some(),
                })
            })
            .collect()
    }
}

pub struct SqlTaskStore {
    pool: PgPool,
}

impl SqlTaskStore {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }
}

#[async_trait]
impl TaskStore for SqlTaskStore {
    async fn list_range(&self, input: ListTasksRangeInput) -> StoreResult<Vec<Task>> {
        let rows = sqlx::query(&format!(
            r#"select {TASK_COLUMNS} from tasks
               where user_id = $1::uuid and deleted_at is null and date >= $2 and date <= $3
               order by date, start_at"#
        ))
        .bind(&input.user_id)
        .bind(&input.from)
        .bind(&input.to)
        .fetch_all(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;
        rows.iter().map(to_task).collect()
    }

    async fn find_by_id(&self, input: &TaskRefInput) -> StoreResult<Option<Task>> {
        let row = sqlx::query(&format!(
            r#"select {TASK_COLUMNS} from tasks
               where user_id = $1::uuid and id = $2::uuid and deleted_at is null"#
        ))
        .bind(&input.user_id)
        .bind(&input.id)
        .fetch_optional(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;
        row.as_ref().map(to_task).transpose()
    }

    async fn list_by_project(&self, input: ListTasksByProjectInput) -> StoreResult<Vec<Task>> {
        let rows = sqlx::query(&format!(
            r#"select {TASK_COLUMNS} from tasks
               where user_id = $1::uuid and project_id = $2::uuid and deleted_at is null
               order by date, start_at"#
        ))
        .bind(&input.user_id)
        .bind(&input.project_id)
        .fetch_all(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;
        rows.iter().map(to_task).collect()
    }

    async fn counts_by_project(&self, user_id: &str) -> StoreResult<Vec<ProjectTaskCounts>> {
        // COUNT은 NULL을 세지 않으므로 done 행만 1로 남겨 센다
        let rows: Vec<(uuid::Uuid, i64, i64)> = sqlx::query_as(
            r#"select project_id, count(*) as total,
                      count(case when status = 'done' then 1 end) as done
               from tasks
               where user_id = $1::uuid and deleted_at is null and project_id is not null
               group by project_id"#,
        )
        .bind(user_id)
        .fetch_all(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;
        Ok(rows
            .into_iter()
            .map(|(project_id, total, done)| ProjectTaskCounts {
                project_id: project_id.to_string(),
                total,
                done,
            })
            .collect())
    }

    async fn create(&self, input: CreateTaskInput) -> StoreResult<Task> {
        let (mut tx, seq) = begin_with_sync_seq(&self.pool, &input.user_id)
            .await
            .map_err(StoreError::infrastructure)?;
        let last: Option<(String,)> = sqlx::query_as(
            "select position from tasks where user_id = $1::uuid order by position desc limit 1",
        )
        .bind(&input.user_id)
        .fetch_optional(&mut *tx)
        .await
        .map_err(StoreError::infrastructure)?;

        let row = sqlx::query(&format!(
            r#"insert into tasks
               (id, user_id, project_id, title, date, start_at, duration_min, status, position,
                sync_seq, deleted_at, completed_at)
               values ($1::uuid, $2::uuid, $3::uuid, $4, $5, $6, $7, 'todo', $8, $9, null, null)
               returning {TASK_COLUMNS}"#
        ))
        .bind(new_id())
        .bind(&input.user_id)
        .bind(&input.project_id)
        .bind(&input.title)
        .bind(&input.date)
        .bind(&input.start_at)
        .bind(input.duration_min as i32)
        .bind(order_key_after(last.map(|(position,)| position).as_deref()))
        .bind(seq as i32)
        .fetch_one(&mut *tx)
        .await
        .map_err(StoreError::infrastructure)?;
        let task = to_task(&row)?;
        tx.commit().await.map_err(StoreError::infrastructure)?;
        Ok(task)
    }

    async fn update(&self, input: UpdateTaskInput) -> StoreResult<Option<Task>> {
        let (mut tx, seq) = begin_with_sync_seq(&self.pool, &input.user_id)
            .await
            .map_err(StoreError::infrastructure)?;
        let now = Utc::now();

        // 지정된 필드만 SET 절에 넣는다 — 필드 단위 LWW라 행 전체 덮어쓰기가 없다.
        // 자리표시자 번호는 아래 bind 순서와 같은 순서로 매긴다.
        let mut columns: Vec<&str> = Vec::new();
        if !input.patch.title.is_absent() {
            columns.push("title");
        }
        if !input.patch.project_id.is_absent() {
            columns.push("project_id");
        }
        if !input.patch.date.is_absent() {
            columns.push("date");
        }
        if !input.patch.start_at.is_absent() {
            columns.push("start_at");
        }
        if !input.patch.duration_min.is_absent() {
            columns.push("duration_min");
        }
        if !input.patch.status.is_absent() {
            // 완료 시각은 상태에 종속된다 — done이면 지금, 아니면 해제
            columns.push("status");
            columns.push("completed_at");
        }
        let assignments: Vec<String> = columns
            .iter()
            .enumerate()
            .map(|(index, column)| format!("{column} = ${}", index + 1))
            .collect();

        let seq_placeholder = columns.len() + 1;
        let updated_at_placeholder = columns.len() + 2;
        let user_placeholder = columns.len() + 3;
        let id_placeholder = columns.len() + 4;

        let sql = format!(
            r#"update tasks set {}{}version = version + 1,
                   updated_at = ${updated_at_placeholder}, sync_seq = ${seq_placeholder}
               where user_id = ${user_placeholder}::uuid and id = ${id_placeholder}::uuid
                 and deleted_at is null
               returning {TASK_COLUMNS}"#,
            assignments.join(", "),
            if assignments.is_empty() { "" } else { ", " },
        );

        let mut query = sqlx::query(&sql);
        if let Patch::Set(title) = &input.patch.title {
            query = query.bind(title);
        }
        if let Patch::Set(project_id) = &input.patch.project_id {
            query = query.bind(project_id.as_ref().and_then(|id| id.parse::<uuid::Uuid>().ok()));
        }
        if let Patch::Set(date) = &input.patch.date {
            query = query.bind(date);
        }
        if let Patch::Set(start_at) = &input.patch.start_at {
            query = query.bind(start_at);
        }
        if let Patch::Set(duration_min) = &input.patch.duration_min {
            query = query.bind(*duration_min as i32);
        }
        if let Patch::Set(status) = &input.patch.status {
            query = query.bind(status.as_str());
            query = query.bind((*status == TaskStatus::Done).then_some(now));
        }
        query = query.bind(seq as i32).bind(now).bind(&input.user_id).bind(&input.id);

        let row = query.fetch_optional(&mut *tx).await.map_err(StoreError::infrastructure)?;
        let task = row.as_ref().map(to_task).transpose()?;
        tx.commit().await.map_err(StoreError::infrastructure)?;
        Ok(task)
    }

    async fn remove(&self, input: &TaskRefInput) -> StoreResult<bool> {
        let (mut tx, seq) = begin_with_sync_seq(&self.pool, &input.user_id)
            .await
            .map_err(StoreError::infrastructure)?;
        let now = Utc::now();
        let row: Option<(uuid::Uuid,)> = sqlx::query_as(
            r#"update tasks set deleted_at = $1, updated_at = $1, sync_seq = $2
               where user_id = $3::uuid and id = $4::uuid and deleted_at is null
               returning id"#,
        )
        .bind(now)
        .bind(seq as i32)
        .bind(&input.user_id)
        .bind(&input.id)
        .fetch_optional(&mut *tx)
        .await
        .map_err(StoreError::infrastructure)?;
        tx.commit().await.map_err(StoreError::infrastructure)?;
        Ok(row.is_some())
    }

    async fn changes_since(&self, input: ListChangesInput) -> StoreResult<Vec<TaskChange>> {
        let rows = sqlx::query(&format!(
            r#"select {TASK_COLUMNS}, sync_seq, deleted_at from tasks
               where user_id = $1::uuid and sync_seq > $2
               order by sync_seq"#
        ))
        .bind(&input.user_id)
        .bind(input.cursor as i32)
        .fetch_all(&self.pool)
        .await
        .map_err(StoreError::infrastructure)?;

        rows.iter()
            .map(|row| {
                let deleted_at: Option<chrono::DateTime<Utc>> =
                    row.try_get("deleted_at").map_err(StoreError::infrastructure)?;
                Ok(TaskChange {
                    task: to_task(row)?,
                    sync_seq: row.try_get::<i32, _>("sync_seq").map_err(StoreError::infrastructure)? as i64,
                    deleted: deleted_at.is_some(),
                })
            })
            .collect()
    }

    async fn sync_cursor(&self, user_id: &str) -> StoreResult<i64> {
        current_sync_seq(&self.pool, user_id).await.map_err(StoreError::infrastructure)
    }
}
